### PR TITLE
Add integration test

### PR DESCRIPTION
**Pull request by kelkawi-a**
This PR adds an integration test to ensure the application can be deployed successfully and a Github action for running integration tests on merging with the `main` branch.

### CODE REVIEW

File: .github/workflows/integration_test.yaml [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+        with:
+          juju-channel: 3.1/stable
+          provider: microk8s
+          microk8s-addons: "ingress storage dns rbac registry"
+          channel: 1.25-strict/stable
+      - name: Run integration tests
+        # set a predictable model name so it can be consumed by charm-logdump-action
+        run: tox -e integration -- --model testing
+      - name: Dump logs
+        uses: canonical/charm-logdump-action@main
```
Comment (kian99) [inline .github/workflows/integration_test.yaml:24]: This action is now available on the [v2.6.0 release](https://github.com/canonical/charming-actions/releases/tag/2.6.0). 

As a side note for the future, I recommend setting up renovatebot to get your actions and deps pinned to exact commits or versions and have renovate update them automatically. 

This next bit is probably worth a CSShare, but the observability charms and identity charms have some nice CI setup to automatically merge renovatebot PRs that pass unit/integration tests.
I've adapted Hydra's [renovate config](https://github.com/canonical/hydra-operator/blob/main/renovate.json) for example in JIMM. Then you can setup [some](https://github.com/canonical/hydra-operator/blob/main/.github/workflows/auto-approver.yaml) cool [automation](https://docs.renovatebot.com/key-concepts/automerge/) to get those PRs merged without user intervention.

File: tests/integration/conftest.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+            APP_NAME_TEMPORAL_SERVER,
+            channel="edge",
+            config={"num-history-shards": 1},
+        ),
+        ops_test.model.deploy(APP_NAME_TEMPORAL_ADMIN, channel="edge"),
+        ops_test.model.deploy("postgresql-k8s", channel="14/stable", trust=True),
+        ops_test.model.deploy("minio", channel="edge"),
+    )
+
+    async with ops_test.fast_forward():
```
Comment (kian99) [inline tests/integration/conftest.py:42]: Just a heads up, I've had issues wrapping "wait_for_idle" with fast_forward. Fast_forward worked well when I used it to wrap the deploy step but what happened with wait_for_idle was specifically with the traefik-k8s charm, it was updating the status every update interval (something like every 5-10s) so it was never "idle" which checks if the charm status hasn't changed in ~15s.

File: tests/integration/test_charm.py [hunk 3]
```diff
@@ -1,32 +1,28 @@
...
+    """Integration tests for charm."""
+
+    async def test_deployment(self, ops_test: OpsTest):
+        url = await get_unit_url(ops_test, application=APP_NAME_AIRBYTE_SERVER, unit=0, port=8001)
+        logger.info("curling app address: %s", url)
+
+        response = requests.get(f"{url}/api/v1/health", timeout=300)
+        print(response.json())
+        assert response.status_code == 200
+        assert response.json().get("available")
```
Comment (kian99) [inline tests/integration/test_charm.py:28]: How do you tell if your worklow stuff ran successfully, I imagine you just want to check here that Airbyte is running though. Also I'm curious why Airbyte and Temporal are needed for this integration test?

**Review by kian99 — APPROVED** [review]

File: tests/integration/helpers.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
+#!/usr/bin/env python3
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Temporal charm integration test helpers."""
```
Comment (AmberCharitos) [inline tests/integration/helpers.py:5]: I guess this is `Airbyte` instead of `Temporal`

File: tests/integration/temporal_client/workflows.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+
+"""Temporal client sample workflow."""
```
Comment (AmberCharitos) [inline tests/integration/temporal_client/workflows.py:5]: Is this sample workflow run just to validate temporal charm is working as expected? Is this testing Airbyte in any way?

**Review by AmberCharitos — COMMENTED** [review]
I agree with Kian's comment - is there a way we can simulate an Airbyte workflow and validate it?

File: tests/integration/helpers.py [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+            "upstream-source"
+        ],
+        "airbyte-cron": METADATA["resources"]["airbyte-cron"]["upstream-source"],
+        "airbyte-pod-sweeper": METADATA["resources"]["airbyte-pod-sweeper"]["upstream-source"],
+        "airbyte-server": METADATA["resources"]["airbyte-server"]["upstream-source"],
+        "airbyte-workers": METADATA["resources"]["airbyte-workers"]["upstream-source"],
+    }
+
+
+async def run_sample_workflow(ops_test: OpsTest):
```
Comment (gtato) [inline tests/integration/helpers.py:44]: what is this for?

File: tests/integration/conftest.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+        )
+        await ops_test.model.wait_for_idle(
+            apps=[APP_NAME_TEMPORAL_SERVER, APP_NAME_TEMPORAL_ADMIN],
+            status="blocked",
+            raise_on_blocked=False,
+            timeout=600,
+        )
+
+        await perform_temporal_integrations(ops_test)
+        await create_default_namespace(ops_test)
```
Comment (gtato) [inline tests/integration/conftest.py:54]: do you have to create the default namespace manually, isn't the relation between temporal and airbyte going to create a namespace?

File: tests/integration/test_charm.py [hunk 8]
```diff
@@ -1,32 +1,77 @@
...
+        job_successful = False
+        for i in range(30):
+            logger.info(f"attempt {i} to get job status")
+            status = check_airbyte_job_status(api_url, job_id)
+
+            if status == "failed":
+                break
+
+            if status == "succeeded":
+                logger.info(f"attempt {i}: job successful!")
```
Comment (gtato) [inline tests/integration/test_charm.py:70]: is this storing pikachu in the db or smth? :smile:

**Review by gtato — APPROVED** [review]
lgtm

File: .github/workflows/integration_test.yaml (hunk 1 above)
Comment (kelkawi-a) [reply .github/workflows/integration_test.yaml]: To accommodate for the new integration test of running an Airbyte sync, I actually had to switch to the integration tests workflow offered by the IS Devops team to be able to use larger self-hosted Github runners.

File: tests/integration/test_charm.py (hunk 3 above)
Comment (kelkawi-a) [reply tests/integration/test_charm.py]: I've added another test to create and run an Airbyte connection.

File: tests/integration/temporal_client/workflows.py (hunk 5 above)
Comment (kelkawi-a) [reply tests/integration/temporal_client/workflows.py:5]: This runs a sample workflow to test that Temporal is up and running, I've added another test to run an Airbyte sync.

File: tests/integration/helpers.py (hunk 6 above)
Comment (kelkawi-a) [reply tests/integration/helpers.py:45]: I realized it wasn't being called so I updated it, but this is to run a sample workflow on Temporal to ensure it's up and running.

File: tests/integration/conftest.py (hunk 7 above)
Comment (kelkawi-a) [reply tests/integration/conftest.py:58]: There's currently no relation between Temporal and Airbyte, so this step needs to be done manually.

File: tests/integration/test_charm.py (hunk 8 above)
Comment (kelkawi-a) [reply tests/integration/test_charm.py]: Yup :D

**Pull request merged**